### PR TITLE
Update TypeScript types, check types against JavaScript tests/examples

### LINKED ISSUE #135: Typescript definitions
The Typescript definitions for this project seem to be lacking. I've seen a PR that has been closed because apparently it was "poorly written" but other than that I can't really find issues regarding types.

I'd love to use this library but right now I'm quite limited with the missing type definitions. I need the definitions for pre-defining questions (prompts?) before actually calling the methods.

### LINKED ISSUE #202: ArrayPromptOptions.choices should be (string | Choice)[], not string[] | Choice[]
> `choices: string[] | Choice[]`

This means TypeScript will not allow you to use an array that contains _both_ `string`s and `Choice`s. 

Did you mean that every item in the array must be `string | Choice`?:

> `choices: (string | Choice)[]

I've verified on the CLI these seem to work in the latter way.

### LINKED ISSUE #212: [types] Module '"../node_modules/enquirer"' has no exported member 'MultiSelect'
hey jon, is there something that broke recently with the types? i cant seem to import MultiSelect. when i import `prompt` instead and use a `type: 'multiselect'`, the `choices` field doesnt seem to accept an array. whats the status? i might be able to contribute a fix but just figured i would get a general take of the state of typing in this lib. i had expected it would be first class...

### LINKED ISSUE #262: Why are properties like AutoComplete not enumerable?
I am new to enquirer. I started by trying to use AutoComplete in a typescript project and found that it wouldn't import with `import { AutoComplete } from 'enquirer'`.

I later found issue #135, so I realise that there are issues with the typescript definitions.

Before finding #135, I experimented with different imports and requires, and console.logged the resultant object. I saw that the AutoComplete property does exist, but is not enumerable.

I am curious, why is AutoComplete not enumerable?

It is part of the public interface of this package.

Is it intentionally made this way because there is some benefit?

### LINKED ISSUE #263: PromptOptions union type does not work well in TypeScript 3.8.x
The following code
```
const { doSomething } = await prompt<{ doSomething: 'Yes' | 'No' }>({
  type: 'autocomplete',
  name: 'doSomething',
  message: 'Some Message',
  choices: ['Yes', 'No'],
  initial: 'Yes'
});
```

Gives the following compile error:
```
Argument of type '{ type: string; name: string; message: string; choices: string[]; initial: string; }' is not assignable to parameter of type 'ArrayPromptOptions | BasePromptOptions | BooleanPromptOptions | StringPromptOptions | ... 4 more ... | (ArrayPromptOptions | ... 6 more ... | ((this: Enquirer<...>) => PromptOptions))[]'.
  Object literal may only specify known properties, and 'choices' does not exist in type 'BasePromptOptions'.ts(2345)
```

The problem is the `type` property on `BasePromptOptions`. This breaks the union type `PromptOptions` so that the compiler does not see that `type` as the discriminant.

Versions:
enquirer: 2.3.2
typescript: 3.8.3
node: 12.12.0
yarn: 1.22.0

### PR DESCRIPTION
**Pull request by bitjson**
Still a work in progress – I'm just trying to get @unional's fantastic work on #261 over the finish line. 

This PR builds on #261 – I've removed the duplicated set of tests (which were being used to develop the types), and instead I'm checking the rest of the repo (`guide`, `examples`, `lib`, `recipes`, `test`) with the new types. I also reverted some formatting changes to keep this PR as thin as possible.

There are ~500 TS issues remaining (many of which were already marked as `TODO` by @unional), but once these issues are fixed, the types should be very complete, and much easier to maintain (`npm test` now validates them). If you're interested in helping, please [send PRs to this branch](https://github.com/bitjson/enquirer/tree/typings), and I'll merge them to add them to this PR.

Closes #82
Closes #135 
Closes #202 
Closes #204 
Closes #212 
Closes #252 
Closes #258 
Closes #261 
Closes #262
Closes #263
Closes #286

### CODE REVIEW

**Comment by bitjson** [comment]
Still fixing issues, but just wanted to highlight an example of where running TypeScript against the library can help to fix easily-missed issues:

```
test/type.string.js:14:36 - error TS2551: Property 'Prompt' does not exist on type 'typeof StringPrompt'. Did you mean 'prompt'?

14       assert(Baz.Prompt === Prompt.Prompt);
                                      ~~~~~~

  lib/prompt.js:417:14
    417   static get prompt() {
                     ~~~~~~
    'prompt' is declared here.
```

The test:
```js

describe('string prompt', function() {
  describe('class', () => {
    it('should expose static method for getting Prompt class', () => {
      class Foo extends Prompt {}
      class Bar extends Foo {}
      class Baz extends Bar {}
      assert(Baz.Prompt === Prompt.Prompt);
    });
  });
```

Right now the test also passes when you `assert(Baz.Prompt === undefined);`. 

Was this a regression from a few years ago? Can we just delete this test now? (The static `prompt` property on each class is unique, so comparing `Baz.prompt === Prompt.prompt` also isn't valuable.)

**Comment by bitjson** [comment]
Ok, 

Hey @unional – I think this is just about ready to merge

**Comment by bitjson** [comment]
Ok, I think this is just about ready to merge. There are definitely still problems, but it's already a huge improvement over the current types. I tried to tone down the changes as much as possible – I'm now only testing the types against a subset of the `examples` directory. (Testing against the rest of the repo will require a lot more work, but may be possible as the types improve.)

I think a good strategy from here is to slowly add new examples to the set of "covered" examples as we continue to improve the types. To add a new example, we just add it to the `include` list in `tsconfig.json`.

I left just 3 examples in the list that still produce errors – @unional if you get a chance, would you mind taking a look at this issue? If we can get this issue solved, I think it will handle around half the remaining type issues in the `examples`.

```
examples/everything.js:16:3 - error TS2345: Argument of type '{ type: string; name: string; message: string; hint: string; limit: number; header: any; format(this: MultiSelect): string; pointer(state: any, choice: any, i: any): string; footer(state: any): string; result(this: MultiSelect, names: any): Record<...>; choices: ({ ...; } | { ...; })[]; }' is not assignable to parameter of type 'Question'.
  Object literal may only specify known properties, and 'limit' does not exist in type 'Question'.

16   limit: 6,
     ~~~~~~~~

examples/log-keypress-example.js:10:3 - error TS2345: Argument of type '{ type: string; name: string; message: string; limit: number; header(): string; pointer(state: any, choice: any, i: any): string; choices: ({ name: string; value: string; } | { name: string; value: string; hint: string; })[]; }' is not assignable to parameter of type 'Question'.
  Object literal may only specify known properties, and 'limit' does not exist in type 'Question'.

10   limit: 5,
     ~~~~~~~~

examples/toggle/prompt.js:7:3 - error TS2345: Argument of type '{ message: string; enabled: string; disabled: string; }' is not assignable to parameter of type 'Question'.
  Object literal may only specify known properties, and 'enabled' does not exist in type 'Question'.

7   enabled: 'Yep',
    ~~~~~~~~~~~~~~


Found 3 errors.
```

I think we may need to add some sort of generic type which allows us to specify some "allowable options" for each prompt type? Adding to the `[Type]Question` types doesn't seem to help here, but I'm also not an expert at writing ambient types. 😅

If we can't get those issues solved soon, we should just skip checking the problematic examples. These new types are already a much better "baseline" for future work, and testing against the JS examples will make maintenance much easier.

Again, nice work on this @unional!

**Comment by unional** [comment]
> I left just 3 examples in the list that still produce errors

Need to figure out where do these two properties apply and add them to the corresponding type. They looks like base `Question` property but in fact it is not.

One problem that makes adding type hard in this library is that a lot of code are pushed to the base `Prompt` class. But the property is actually not used universally. i.e., the code is breaking some SOLID principles 😆 .

Checking it against the JavaScript example might not work well because it is more flexible and allow these extra properties exist in the base class.

What I mean is, this will come down to a trade-off. If we want to work well with `checkJS`, we have to give up some soundness of the type.
If the type is intended to only support TypeScript (and `allowjs` as the TypeScript language service will not punish your JavaScript code in that mode), we can make the type better by doing what I suggest here.

**Comment by bitjson** [comment]
I think I've figured out the issue with those errors – just doing a bit more polishing and this PR should be ready to go.

> Checking it against the JavaScript example might not work well because it is more flexible and allow these extra properties exist in the base class.
>
> What I mean is, this will come down to a trade-off. If we want to work well with checkJS, we have to give up some soundness of the type.

I fought with trying to get types over the `test` directory for a while, so I definitely appreciate how hard it is to cover the 
deeply-inherited-JS architecture style with types. 😄

I think the best move is to provide types which are loose enough to work for all examples – they should basically document the "user-facing API", but we don't need to be too concerned about re-specifying the library internals in TypeScript. I think what you've done is already **very close** to that goal: we just need to allow a few more options in several question types. 

I really, really appreciate the `Question` discriminated union type you've put together – that makes the types much more useful than a generic top-level Question would be. I don't think a looser type would be "more correct", e.g. if you set a `pointer` method on a boolean prompt, the library won't throw an error, but it also won't do anything. So the types don't need to "support" that. (If there's a TypeScript user out there who wants to monkey patch it in, it's reasonable for us to expect them to extend the type definition too.)

TLDR: I think these types are already appropriately loose and a major improvement  🚀

**Comment by bitjson** [comment]
Ok, I cleaned up a few more issues – there are still places we need to improve, but I think we're in a good spot to make smaller, iterative changes now. (Especially now that we're testing against the actual JS examples.)

@jonschlinkert would you be willing to merge these TypeScript changes? Is there anything else you'd like me change?

File: .travis.yml [hunk 1]
```diff
@@ -9,3 +9,5 @@ node_js:
   - '11'
   - '10'
   - '8'
+before_script:
+  - cd examples && npm install && cd ../guide && npm install && cd ..
```
Comment (bitjson) [inline .travis.yml:13]: Some of the examples use external dependencies which aren't in the top-level `package.json`, so we have to install them before running `tsc` in Travis. (Otherwise TypeScript throws errors for missing dependencies.)

**Comment by bitjson** [comment]
(I force-pushed to retry CI. [Tests on Node.js 8 fail randomly](https://travis-ci.org/github/enquirer/enquirer/jobs/715950386) – not related to this PR, which doesn't touch any JS.)

File: .travis.yml (hunk 1 above)
Comment (jonschlinkert) [reply .travis.yml:13]: thanks!

**Review by jonschlinkert — COMMENTED** [review]
> the code is breaking some SOLID principles

Yeah, there are tradeoffs. If you've seen any of my other libraries, I'm a big fan of SOLID principles, I tend to favor responsibility libraries. IMHO, in Enquirer the benefits of inheritance far outweigh the advantages of SOLID, from the standpoints of both maintenance and implementation. It's just too much work to redefine everything on every class, and repeated code everywhere lends itself to inconsistencies when minor improvements or tweaks are made. 

> (I force-pushed to retry CI. Tests on Node.js 8 fail randomly – not related to this PR, which doesn't touch any JS.)

I think it's related to a dev dependency (time-require I think). Don't worry about it, I'll get that fixed. If that is indeed what's causing it, I'll either fix the node version inconsistency or just remove that unnecessary dev dep.

**Comment by bitjson** [comment]
@jonschlinkert thanks for taking a look! Please let me know if there's anything else you'd like me to change before merging this. 🚀

I'm current using a version of this PR in my development branch of [Bitauth CLI](https://github.com/bitauth/bitauth-cli), where I'm continuing to improve the types as needed. 

After this PR is merged, future PRs should have much smaller diffs and only increase "TypeScript coverage".  (In this PR, [only a few examples](https://github.com/enquirer/enquirer/blob/4ca593eb04efaacfb1c2715ce2ec4e74ab49ba40/tsconfig.json#L16-L26) are "covered", but we can slowly expand that list until it includes everything in the `examples`, `guide`, and `test` directories.)

**Comment by linzhiq** [comment]
^ any update on this 😬

**Comment by KamalAman** [comment]
@jonschlinkert / @doowb  🙏  💖 🙏 Getting this in would be very much appreciated 🙏  💖 🙏

**Comment by kherock** [comment]
👍 Just confirming that these changes work great in a CLI project I'm writing in TS, would really appreciate if these changes made in in soon!

**Comment by bitjson** [comment]
Hey @jonschlinkert and @doowb, I think this has been well-reviewed by some of the above commenters, and the changes are contained to only TypeScript-related parts of the project. (It also resolves/closes 11 other issues and PRs! 😄)

I think merging is pretty low-risk now: would it be possible to get this released soon?

**Comment by eleith** [comment]
as a workaround till this lands, you can update your `tsconfig.json` to use the types provided by this branch as follows (the below example assumes you download `index.d.ts` from this branch to `types/enquirer.d.ts`)

```json
{                                                                                                                                                                                                                
   "compilerOptions": {                                                                                                                                                                                                                                                                                                                                                                                   
     "paths": {                                                                                                                                                                                                   
       "enquirer": ["./types/enquirer.d.ts"]                                                                                                                                                                      
     }                                                                                                                                                                                                            
   }                                                                                                                                                                                                                                                                                                                                                                                          
}  
```

**Comment by ericvera** [comment]
> as a workaround till this lands, you can update your `tsconfig.json` to use the types provided by this branch as follows (the below example assumes you download `index.d.ts` from this branch to `types/enquirer.d.ts`)
> 
> ```json
> {                                                                                                                                                                                                                
>    "compilerOptions": {                                                                                                                                                                                                                                                                                                                                                                                   
>      "paths": {                                                                                                                                                                                                   
>        "enquirer": ["./types/enquirer.d.ts"]                                                                                                                                                                      
>      }                                                                                                                                                                                                            
>    }                                                                                                                                                                                                                                                                                                                                                                                          
> }  
> ```

Thanks @eleith for the workaround! In case someone runs into issues with the solution you also need to set `baseUrl` for this `paths` work.

 ```json
{
  "compilerOptions": {
    "baseUrl": ".",
    "paths": {
      "enquirer": ["./types/enquirer.d.ts"]
    }
  }                                                                                                                                                                                                                                                                                                                                                                              }  
```

**Comment by youbek** [comment]
Any updates on this? Thanks :)

**Comment by lenovouser** [comment]
I've tested this as well, seems to work flawlessly so far. Can this be merged @jonschlinkert and @doowb?

File: index.d.ts [hunk 2]
```diff
@@ -138,14 +69,1128 @@ declare class Enquirer<T = object> extends EventEmitter {
...
+        ) =>
+          | SelectQuestion.ChoiceOptions[]
+          | Promise<SelectQuestion.ChoiceOptions[]>;
+      } & types.QuestionBase;
+
+      export type SingleOptions = { multiple?: false } & Type &
+        Options &
+        types.Initializer<string | number, string> &
+        types.Formatter<string, string>;
+      export type MultipleOptions = { multiple: true } & Type &
```
Comment (ProdigySim) [inline index.d.ts:173]: This `Type &` seems superfluous, it resolves to nothing in my local setup.

**Comment by ProdigySim** [comment]
A couple enhancements that would be nice to have (for a V2): `prompt(question)` should have its `Answer` type inferred from the question input. This is a trickier task, but I've got some initial work here that accomplishes it for a few of the supported prompt types:

https://gist.github.com/ProdigySim/73913e526d12719476582238a3fce2aa

Supports:
- [x] input
- [x] numeral
- [x] invisible
- [x] password
- [x] list
- [x] autocomplete

Missing:
- [ ] select
- [ ] multiselect
- [ ] quiz
- [ ] ???

**Comment by rebeccahum** [comment]
Hi! We are seeing that the type definition for `ConfirmQuestionOptions` is expecting a string for the `initial` property, however, it should allow a boolean per https://github.com/enquirer/enquirer/blob/70bdb0fedc3ed355d9d8fe4f00ac9b3874f94f61/lib/prompts/confirm.js#L8

I think we need to change `initial?: T | ( () => Promise< T > | T );` to `initial?: T | ( () => Promise< T > | T ) | boolean | ( () => Promise< boolean > | boolean );` under `export type Question< T extends types.Answer = string, P extends Prompt< T > = Prompt< T > > = {`

Thankssss

**Comment by Hotell** [comment]
any updates on this ? is this planned to get in ? cheers
